strings/base_array.h

```diff
@@ -31,6 +31,20 @@ WINRT_EXPORT namespace winrt
             array_view(value.begin(), static_cast<size_type>(value.size()))
         {}
 
+#ifdef __cpp_lib_span
+        template <typename C, std::size_t extent>
+        array_view(std::span<C, extent> span) noexcept :
+            array_view(span.data(), static_cast<size_type>(span.size())) 
+        {
+            WINRT_ASSERT(span.size() <= UINT_MAX);
```

Comment:
There are other places (like the constructor of hstring) where currently no check is done. Need to also add this assertion there. Another thing, I would probably recommend using WINRT_VERIFY instead as WINRT_ASSERT is removed in release mode.

Reply:
Looks like this should have been an `if (too large) { throw invalid arg }` instead. That's what other places do.

Reply:
But then there's the other places in array_view that are more blase about it:

```c++
        array_view(pointer first, pointer last) noexcept :
            m_data(first),
            m_size(static_cast<size_type>(last - first))
        {}

        array_view(std::initializer_list<value_type> value) noexcept :
            array_view(value.begin(), static_cast<size_type>(value.size()))
        {}
```

@kennykerr what's the overall philosophy here? While it's _super unlikely_ that anybody passes UINT_MAX elements to any WinRT API, I suppose it's not unreasonable...

Reply:
(And now you've got me worried about truncation errors in string comparisons and other things...)

Reply:
C++/WinRT will mostly (debug) assert for things like this. The primary outlier is when an HSTRING initializer is not null terminated in which case it panics. 

I'm not opposed to turning these into exceptions but that would ideally be done consistently throughout the library and consideration be taken about code size impact.

Reply:
The issue of returning span for a local vector could be solved by deleting the rvalue qualified conversion operator but unfortunately, the standard library does not use this mechanism anywhere.

Reply:
Tracked in #1347 so we can make progress here.